### PR TITLE
Add Monkey C and Win32 Message File languages

### LINKED ISSUE #5563: Add support for Monkey C language
## Language name
Monkey C

## URL of example repository
https://github.com/Wheellog/Companion.Garmin

## URL of syntax highlighting grammar

https://developer.garmin.com/connect-iq/monkey-c/functions/

## Most popular extensions

https://marketplace.visualstudio.com/items?itemName=garmin.monkey-c

## Detected language

.mc

### PR DESCRIPTION
**Pull request by Nixinova**
- Resolve #5563

## Description

1. Adds the MonkeyC language.
  Color from [logo](https://developer.garmin.com/connect-iq/resources/programmers-guide/smart-monkey.png) on [website](https://developer.garmin.com/connect-iq/monkey-c/)

2. Adds the [Windows Message File](https://docs.microsoft.com/en-us/windows/win32/eventlog/message-text-files) language used by the Win32 API.

3. Adds [Sendmail](https://tldp.org/LDP/nag2/x14661.html)-[files](https://www.oreilly.com/library/view/sendmail-3rd-edition/1565928393/ch04s02.html) to the M4 language

## Checklist
- [x] **I am adding a new language.**
  - [x] The extension of the new language is used in hundreds of repositories on GitHub.com.
    - Search results for each extension:
        - [58K](https://github.com/search?q=extension%3Amc) overall
      - MonkeyC:
        -  [6K+](https://github.com/search?q=extension%3Amc+using+toybox) with exact keywords ("using Toybox", the stdlib)
        - [16K+](https://github.com/search?q=extension%3Amc+using) with just "using"
      - Win32 Messages:
        - [14K](https://github.com/search?q=extension%3Amc+%22MessageId%22&type=code)  with keyword "MessageID"
      - M4 Sendmail:
        - [8.8K](https://github.com/search?q=extension%3Amc+%22divert%22+%22dnl%22&type=code)  with keywords "divert"/"dnl"
  - [x] I have included a real-world usage sample for all extensions added in this PR:
    - MonkeyC:
      - [Sample](https://github.com/uschi2000/garmin-max-speed/blob/develop/source/StartTimer.mc) under [Apache2](https://github.com/uschi2000/garmin-max-speed/blob/develop/LICENSE)
    - Win32 Message File:
      - [Sample](https://github.com/junhuac/equic/blob/9d2dde09689246cd61a5e7bfdc8ca37884043a9c/erts/etc/win32/erlsrv/erlsrv_logmess.mc) under [Apache2](https://github.com/junhuac/equic/blob/master/LICENSE.txt) 
    - M4 Sendmail:
      - [Sample](https://raw.githubusercontent.com/centminmod/centminmod/96caa910e144a41d4747a218e23b0519907564a8/config/sendmail/sendmail.mc) under [gpl3](https://github.com/centminmod/centminmod/blob/master/license.txt)
  
  - [x] I have included a syntax highlighting grammar:
   -  MonkeyC:
      - [Source](https://github.com/ghisguth/vscode-monkey-c)
      - [Preview][monkeyc-preview]
   - Win32 Message File:
     - Source: [NovaGrammars](https://github.com/nixinova/novagrammars)
     - [Preview][win32-preview]
   - M4 Sendmail:
     - uses M4's syntax highlighting

[monkeyc-preview]: https://github-lightshow.herokuapp.com/?utf8=%E2%9C%93&scope=from-url&grammar_format=auto&grammar_url=https%3A%2F%2Fgithub.com%2Fghisguth%2Fvscode-monkey-c%2Fblob%2Fmaster%2Fsyntaxes%2Fmonkeyc.tmLanguage.json&grammar_text=&code_source=from-url&code_url=https%3A%2F%2Fgithub.com%2Fuschi2000%2Fgarmin-max-speed%2Fblob%2Fdevelop%2Fsource%2FStartTimer.mc&code=

[win32-preview]: https://github-lightshow.herokuapp.com/?utf8=y&scope=from-url&grammar_url=https://github.com/Nixinova/NovaGrammars/blob/main/grammars/win32-messages.yaml-tmLanguage&code_source=from-url&code_url=https://github.com/Nixinova/NovaGrammars/blob/main/samples/messages.mc

### CODE REVIEW

File: lib/linguist/languages.yml [hunk 1]
```diff
@@ -3781,6 +3781,16 @@ Monkey:
   ace_mode: text
   tm_scope: source.monkey
   language_id: 236
+Monkey C:
+  type: programming
+  color: "#8D6747"
+  extensions:
+  - ".mc"
```
Comment (lildude) [inline lib/linguist/languages.yml:3788]: There appear to be a lot of other users of the `.mc` extension that don't all appear to be Monkey C and Linguist doesn't know about this extension yet, so they're all going to be classified as Monkey C. I think we're going to need to identify the language of at least one other user of this extension and add it, possibly with a heuristic too.

**Comment by Nixinova** [comment]
Other stuff with `.mc`:
- [14K](https://github.com/search?q=extension%3Amc+%22MessageId%22&type=code) of [th](https://github.com/dotnet/aspnetcore/blob/fdbe28e9c3e6e5ff3f12a9ac75e6f4efc5bd82cc/src/Servers/IIS/AspNetCoreModuleV2/CommonLib/aspnetcore_msg.mc)[is](https://github.com/google/grr/blob/7c9ac0c9cc793cdd696a612ccad938ccf57baa02/grr/client/grr_response_client/nanny/GRRNanny.mc) translation file thing
- [8K](https://github.com/search?q=extension%3Amc+%22divert%22&type=code)/[9K](https://github.com/search?q=extension%3Amc+%22dnl%22&type=code) for some [mail thing](https://github.com/ulli-kroll/gentoo-arm64/blob/042c2f870527892d8053803cfd73741ded76da05/mail-mta/sendmail/files/sendmail-procmail.mc)
- [3K](https://github.com/search?q=extension%3Amc+%22oracle%22&type=code) of something Oracle-related
- [25K](https://github.com/search?q=extension%3Amc+-MessageId+-toybox+-divert+-oracle&type=code) other

**Comment by Nixinova** [comment]
Should be good now.
The 3K "&copy; Oracle" files seem to be assembly but look like they are all inside a `jdk-*` folder.
I can't find any other common languages in the other 25K results.

File: lib/linguist/heuristics.yml [hunk 2]
```diff
@@ -360,6 +360,14 @@ disambiguations:
   rules:
   - language: Unity3D Asset
     pattern: 'tag:unity3d.com'
+- extensions: ['.mc']
+  rules:
+  - language: Win32 Message File
+    pattern: '(?i)MessageId=|^;'
```
Comment (Alhadis) [inline lib/linguist/heuristics.yml:366]: I'd restrict this pattern so `MessageId=` only matches at the beginning of lines (possibly with `/*` sandwiched in-between):

```suggestion
    pattern: '(?i)^[ \t]*(?>\/\*\s*)?MessageId=|^;'
```

As for the `^;` pattern… recalling that M4 is a language-agnostic templating engine, it's possible that `;` might match in less common scenarios (say, when generating the source of a configuration file). I'd therefore match `M4`'s heuristic first, before `Win32 Message Files`. See below.

File: lib/linguist/heuristics.yml [hunk 3]
```diff
@@ -360,6 +360,14 @@ disambiguations:
...
+  - language: M4
+    pattern: '^dnl|^divert\(|^\w+\(`'
```
Comment (Alhadis) [inline lib/linguist/heuristics.yml:368]:
```suggestion
    pattern: '^dnl|^divert\((?:-?\d+)?\)|^\w+\(`[^\n]*?'[),]'
```

1. IIRC, M4's [`divert()` built-in](https://web.mit.edu/gnu/doc/html/m4_9.html#SEC45) only accepts a numeric argument.

2. Traditional implementations of M4 use [\`old-school typewriter quotes'](https://github.com/Alhadis/Atom-XTermQuotes) for delimiting macro arguments.

File: lib/linguist/heuristics.yml [hunk 4]
```diff
@@ -360,6 +360,14 @@ disambiguations:
...
+  - language: Monkey C
+    pattern: '\b(?:using|module|function|class|var)\b\s+\w'
```
Comment (Alhadis) [inline lib/linguist/heuristics.yml:370]: The second `\b` is redundant, since `\s+` is already a word boundary.

**Review by Alhadis — CHANGES_REQUESTED** [review]
Just a few nits picked up during cursory review.

**Comment by Nixinova** [comment]
The syntax error was added in the second edit but for some reason the first edit failed as well. Whats up with that.

**Review by Alhadis — APPROVED** [review]

File: lib/linguist/heuristics.yml [hunk 5]
```diff
@@ -365,7 +365,7 @@ disambiguations:
   - language: Win32 Message File
     pattern: '(?i)^[ \t]*(?>\/\*\s*)?MessageId=|^;'
   - language: M4
-    pattern: '^dnl|^divert\((?:-?\d+)?\)|^\w+\(`[^\n]*?'[),]'
+    pattern: '^dnl|^divert\((?:-?\d+)?\)|^\w+\(`[^\n]*?''[),]'
```
Comment (Alhadis) [inline lib/linguist/heuristics.yml:368]: I swear I can never remember what YAML's frikkin' syntax is for escaping single-quotes… 🤦‍♂️

**Review by lildude — APPROVED** [review]

**Pull request merged**